### PR TITLE
Add tests and fixes for sites with no species

### LINKED ISSUE #39: Error when site contains no species in fd_fdiv() and fd_feve()
Don't know how this bug slip in the cracks but here it is:
``` r
fundiversity::fd_fdiv(fundiversity::traits_plants, fundiversity::site_sp_plants)
#> Error in traits[names(sub_site), , drop = FALSE]: indice hors limites

sum(fundiversity::site_sp_plants[10,])
#> [1] 0

fundiversity::fd_fdiv(fundiversity::traits_plants,
                      fundiversity::site_sp_plants[1:9,])
#>        site      FDiv
#> 1  elev_250 0.6341541
#> 2  elev_500 0.6543063
#> 3 elev_1000 0.7111319
#> 4 elev_1500 0.7546447
#> 5 elev_2000 0.7437969
#> 6 elev_2500 0.7620128
#> 7 elev_3000 0.6939043
#> 8 elev_3500 0.6414894
#> 9 elev_3750 0.5879492
```

<sup>Created on 2021-08-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

This should mean that an early check in the individual fd_div() computation should be if all abundance are 0.

### PR DESCRIPTION
**Pull request by Rekyt**
Should fix #39

### CODE REVIEW

File: tests/testthat/test-fdiv.R [hunk 1]
```diff
@@ -15,6 +15,13 @@ test_that("Functional Divergence output format", {
   expect_equal(fdiv$FDiv, 0.7282172, tolerance = 1e-7)
 })
 
+test_that("Functional Divergence works even on sites with no species", {
+  data("traits_plants")
+  data("site_sp_plants")
+
+  fdiv <- expect_silent(fd_fdiv(traits_plants, site_sp_plants))
```
Comment (Bisaloo) [inline tests/testthat/test-fdiv.R:22]: We should have stricter tests than just checking if there is an output. We should check that the output of the relevant line is relevant what we want.

**Review by Bisaloo — COMMENTED** [review]
Do we actually want `NA_real_` when the site is empty? Shouldn't it be 0?

If we want 0, this case should also be handled earlier in the script to avoid pooling multiple sources of problem (empty site vs presence of NA in the site matrix). We should test if `rowSums(sp_com) == 0` when we standardize abundance.

File: tests/testthat/test-fdiv.R (hunk 1 above)
Comment (Rekyt) [reply tests/testthat/test-fdiv.R:22]: Indeed, this test is too basic

**Comment by Rekyt** [comment]
> Do we actually want `NA_real_` when the site is empty? Shouldn't it be 0?

Good question. For me, if the site has no species, it has by definition no functional diversity. But it's true that we can say that it has a richness of 0. So a functional divergence/evenness of 0 could probably make sense. Let met check what other packages do in this case.

> If we want 0, this case should also be handled earlier in the script to avoid pooling multiple sources of problem (empty site vs presence of NA in the site matrix). We should test if `rowSums(sp_com) == 0` when we standardize abundance.

Yes I understand that we should limit the amount of branching in the code due to handling specific cases. As wasn't sure of the best way to proceed.

**Comment by Rekyt** [comment]
`FD::dbFD()` errors when using site with zero-sum abundance:
``` r
FD::dbFD(traits_plants, site_sp_plants, calc.FRic = TRUE, calc.FDiv = TRUE)
#> Error in FD::dbFD(traits_plants, site_sp_plants, calc.FRic = TRUE, calc.FDiv = TRUE): objet 'traits_plants' introuvable
```

<sup>Created on 2021-08-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

For FRic we output a value of `NA` because we can't define a volume with no points, right?

``` r
fundiversity::fd_fric(fundiversity::traits_plants, fundiversity::site_sp_plants[10,,drop = FALSE])
#>        site FRic
#> 1 elev_4000   NA
```

<sup>Created on 2021-08-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>


For Functional Evenness (FEve), it needs a minimum spanning tree. Which has no definition with less than 3 species. So below that its value should be NA, right?

For Functional Divergence, it's a weoghted distance to the centroid, so it could be 0, but what could be the definition of a centroid with no species?

**Comment by Bisaloo** [comment]
My view would be that all indices should evaluate to 0 when there are no species.

FRic is a somewhat different case in my opinion. When there is between 1 and n_traits species, the functional richness exists and is non-zero but we cannot compute it. It should still evaluate to 0 when there are no species.

I don't feel super strongly about this though. You are the expert on the matter and I trust your decision.

**Comment by Rekyt** [comment]
Agree with the use of 0 when there are no species, but I'm puzzled that other packages don't do the same. I'm going to check to make sure how they consider this cases.

> FRic is a somewhat different case in my opinion. When there is between 1 and n_traits species, the functional richness exists and is non-zero but we cannot compute it. It should still evaluate to 0 when there are no species.

Just for FRic I wonder about the definition, what's the volume (in 3D) of a 2D line? It's not defined no? Or is it equal to 0?

See https://www.quora.com/Do-all-2D-shapes-have-no-volume but https://en.wikipedia.org/wiki/Space-filling_curve

**Comment by Bisaloo** [comment]
> Just for FRic I wonder about the definition, what's the volume (in 3D) of a 2D line? It's not defined no? Or is it equal to 0?

From a mathematical point of view, I think it's 0 but in the case of FRic, it makes more sense to use NA in my opinion.

**Comment by Rekyt** [comment]
FYI this is what I found (some pkgs output errors, some 0, some NAs):
``` r
## adiv
adiv::QE(fundiversity::site_sp_plants, dist(fundiversity::traits_plants))
#> Registered S3 method overwritten by 'spdep':
#>   method   from
#>   plot.mst ape
#> Registered S3 methods overwritten by 'adegraphics':
#>   method         from
#>   biplot.dudi    ade4
#>   kplot.foucart  ade4
#>   kplot.mcoa     ade4
#>   kplot.mfa      ade4
#>   kplot.pta      ade4
#>   kplot.sepan    ade4
#>   kplot.statis   ade4
#>   scatter.coa    ade4
#>   scatter.dudi   ade4
#>   scatter.nipals ade4
#>   scatter.pco    ade4
#>   score.acm      ade4
#>   score.mix      ade4
#>   score.pca      ade4
#>   screeplot.dudi ade4
#>           diversity
#> elev_250  3270.0234
#> elev_500  2920.7386
#> elev_1000 2514.5801
#> elev_1500 3027.7986
#> elev_2000 2031.0098
#> elev_2500  959.6946
#> elev_3000  719.9246
#> elev_3500  593.9039
#> elev_3750  494.5225
#> elev_4000    0.0000

## BAT
# dispersion
BAT::dispersion(fundiversity::site_sp_plants, fundiversity::traits_plants)
#>           Dispersion
#> elev_250   0.2898571
#> elev_500   0.2747371
#> elev_1000  0.2466034
#> elev_1500  0.2402311
#> elev_2000  0.2051801
#> elev_2500  0.1644256
#> elev_3000  0.1451699
#> elev_3500  0.1330958
#> elev_3750  0.1248028
#> elev_4000  0.0000000

# evenness
BAT::evenness(fundiversity::site_sp_plants, fundiversity::traits_plants)
#> Error in hclust(as.dist(thisTree)): must have n >= 2 objects to cluster

# rao
BAT::rao(fundiversity::site_sp_plants, distance = fundiversity::traits_plants)
#>                  Rao
#> elev_250  0.07576861
#> elev_500  0.07173550
#> elev_1000 0.06341158
#> elev_1500 0.06105123
#> elev_2000 0.05120676
#> elev_2500 0.04237349
#> elev_3000 0.03669180
#> elev_3500 0.03191835
#> elev_3750 0.03051572
#> elev_4000        NaN

# richness
BAT::alpha(fundiversity::site_sp_plants, fundiversity::traits_plants)
#>            Richness
#> elev_250   9.714367
#> elev_500  10.080198
#> elev_1000  8.655776
#> elev_1500  7.565385
#> elev_2000  6.082294
#> elev_2500  4.636093
#> elev_3000  3.336165
#> elev_3500  2.363547
#> elev_3750  1.595673
#> elev_4000  0.000000

## hillR
# with FDis
hillR::hill_func(fundiversity::site_sp_plants, fundiversity::traits_plants)
#> Error in FD::fdisp(d = dij, a = as.matrix(comm)): At least one community has zero-sum abundances (no species).

# without FDis
hillR::hill_func(fundiversity::site_sp_plants, fundiversity::traits_plants, fdis = FALSE)
#> Error in dij[names(df2), names(df2)]: indice hors limites

## mFD
# FEve
mFD::alpha.fd.multidim(fundiversity::traits_plants, fundiversity::site_sp_plants, ind_vect = c("feve"))
#> Warning in check.asb.sp.w(asb_sp_w): Some species are absent from all
#> assemblages.
#> Warning in check.asb.sp.w(asb_sp_w): Some assemblages do not contain species.
#> Warning in check.asb.sp.w(asb_sp_w): Some species are absent from all
#> assemblages.
#> Warning in check.asb.sp.w(asb_sp_w): Some assemblages do not contain species.
#> Registered S3 method overwritten by 'dendextend':
#>   method     from 
#>   rev.hclust vegan
#> [1] "elev_250 done 10%"
#> [1] "elev_500 done 20%"
#> [1] "elev_1000 done 30%"
#> [1] "elev_1500 done 40%"
#> [1] "elev_2000 done 50%"
#> [1] "elev_2500 done 60%"
#> [1] "elev_3000 done 70%"
#> [1] "elev_3500 done 80%"
#> [1] "elev_3750 done 90%"
#> Error in FUN(X[[i]], ...): seulement défini pour des tableaux de données dont toutes les variables sont numériques

# FDiv
mFD::alpha.fd.multidim(fundiversity::traits_plants, fundiversity::site_sp_plants, ind_vect = c("fdiv"))
#> Warning in check.asb.sp.w(asb_sp_w): Some species are absent from all
#> assemblages.

#> Warning in check.asb.sp.w(asb_sp_w): Some assemblages do not contain species.
#> Warning in check.asb.sp.w(asb_sp_w): Some species are absent from all
#> assemblages.
#> Warning in check.asb.sp.w(asb_sp_w): Some assemblages do not contain species.
#> [1] "elev_250 done 10%"
#> [1] "elev_500 done 20%"
#> [1] "elev_1000 done 30%"
#> [1] "elev_1500 done 40%"
#> [1] "elev_2000 done 50%"
#> [1] "elev_2500 done 60%"
#> [1] "elev_3000 done 70%"
#> [1] "elev_3500 done 80%"
#> [1] "elev_3750 done 90%"
#> Error in FUN(X[[i]], ...): seulement défini pour des tableaux de données dont toutes les variables sont numériques

# FRic
mFD::alpha.fd.multidim(fundiversity::traits_plants, fundiversity::site_sp_plants, ind_vect = c("fric"))
#> Warning in check.asb.sp.w(asb_sp_w): Some species are absent from all
#> assemblages.

#> Warning in check.asb.sp.w(asb_sp_w): Some assemblages do not contain species.
#> Warning in check.asb.sp.w(asb_sp_w): Some species are absent from all
#> assemblages.
#> Warning in check.asb.sp.w(asb_sp_w): Some assemblages do not contain species.
#> [1] "elev_250 done 10%"
#> [1] "elev_500 done 20%"
#> [1] "elev_1000 done 30%"
#> [1] "elev_1500 done 40%"
#> [1] "elev_2000 done 50%"
#> [1] "elev_2500 done 60%"
#> [1] "elev_3000 done 70%"
#> [1] "elev_3500 done 80%"
#> [1] "elev_3750 done 90%"
#> Error in FUN(X[[i]], ...): seulement défini pour des tableaux de données dont toutes les variables sont numériques
```

<sup>Created on 2021-08-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.3 (2020-10-10)
#>  os       macOS High Sierra 10.13.6   
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  fr_FR.UTF-8                 
#>  ctype    fr_FR.UTF-8                 
#>  tz       Europe/Paris                
#>  date     2021-08-17                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package           * version    date       lib
#>  abind               1.4-5      2016-07-21 [1]
#>  ade4                1.7-17     2021-06-17 [1]
#>  adegenet            2.1.4      2021-07-17 [1]
#>  adegraphics         1.0-15     2018-12-18 [1]
#>  adephylo            1.1-11     2017-12-18 [1]
#>  adiv                2.0.1      2020-08-26 [1]
#>  ape                 5.5        2021-04-25 [1]
#>  assertthat          0.2.1      2019-03-21 [1]
#>  backports           1.2.1      2020-12-09 [1]
#>  BAT                 2.6.1      2021-07-08 [1]
#>  boot                1.3-28     2021-05-03 [1]
#>  cachem              1.0.5      2021-05-15 [1]
#>  class               7.3-19     2021-05-03 [1]
#>  classInt            0.4-3      2020-04-07 [1]
#>  cli                 3.0.1      2021-07-17 [1]
#>  cluster             2.1.2      2021-04-17 [1]
#>  clusterGeneration   1.3.7      2020-12-15 [1]
#>  coda                0.19-4     2020-09-30 [1]
#>  codetools           0.2-18     2020-11-04 [1]
#>  colorspace          2.0-2      2021-06-24 [1]
#>  combinat            0.0-8      2012-10-29 [1]
#>  crayon              1.4.1      2021-02-08 [1]
#>  crosstalk           1.1.1      2021-01-12 [1]
#>  data.table          1.14.0     2021-02-21 [1]
#>  DBI                 1.1.1      2021-01-15 [1]
#>  deldir              0.2-10     2021-02-16 [1]
#>  dendextend          1.15.1     2021-05-08 [1]
#>  digest              0.6.27     2020-10-24 [1]
#>  dplyr               1.0.7      2021-06-18 [1]
#>  DT                  0.18       2021-04-14 [1]
#>  e1071               1.7-7      2021-05-23 [1]
#>  ellipsis            0.3.2      2021-04-29 [1]
#>  evaluate            0.14       2019-05-28 [1]
#>  expm                0.999-6    2021-01-13 [1]
#>  FactoMineR          2.4        2020-12-11 [1]
#>  fansi               0.5.0      2021-05-25 [1]
#>  fastcluster         1.2.3      2021-05-24 [1]
#>  fastmap             1.1.0      2021-01-25 [1]
#>  fastmatch           1.1-3      2021-07-23 [1]
#>  FD                  1.0-12     2014-08-19 [1]
#>  flashClust          1.01-2     2012-08-21 [1]
#>  fs                  1.5.0      2020-07-31 [1]
#>  fundiversity        0.2.0      2021-05-13 [1]
#>  future              1.21.0     2020-12-10 [1]
#>  future.apply        1.7.0      2021-01-04 [1]
#>  gdata               2.18.0     2017-06-06 [1]
#>  generics            0.1.0      2020-10-31 [1]
#>  geometry            0.4.5      2019-12-04 [1]
#>  ggplot2             3.3.5      2021-06-25 [1]
#>  ggrepel             0.9.1      2021-01-15 [1]
#>  globals             0.14.0     2020-11-22 [1]
#>  glue                1.4.2      2020-08-27 [1]
#>  gmodels             2.18.1     2018-06-25 [1]
#>  gridExtra           2.3        2017-09-09 [1]
#>  gtable              0.3.0      2019-03-25 [1]
#>  gtools              3.9.2      2021-06-06 [1]
#>  highr               0.9        2021-04-16 [1]
#>  hillR               0.5.1      2021-03-02 [1]
#>  hms                 1.1.0      2021-05-17 [1]
#>  htmltools           0.5.1.1    2021-01-22 [1]
#>  htmlwidgets         1.5.3      2020-12-10 [1]
#>  httpuv              1.6.1      2021-05-07 [1]
#>  httr                1.4.2      2020-07-20 [1]
#>  hypervolume         2.0.12     2019-12-06 [1]
#>  igraph              1.2.6      2020-10-06 [1]
#>  jpeg                0.1-9      2021-07-24 [1]
#>  jsonlite            1.7.2      2020-12-09 [1]
#>  KernSmooth          2.23-20    2021-05-03 [1]
#>  knitr               1.33       2021-04-24 [1]
#>  ks                  1.13.2     2021-07-01 [1]
#>  later               1.2.0      2021-04-23 [1]
#>  lattice             0.20-44    2021-05-02 [1]
#>  latticeExtra        0.6-29     2019-12-19 [1]
#>  lazyeval            0.2.2      2019-03-15 [1]
#>  leaps               3.1        2020-01-16 [1]
#>  LearnBayes          2.15.1     2018-03-18 [1]
#>  lifecycle           1.0.0      2021-02-15 [1]
#>  listenv             0.8.0      2019-12-05 [1]
#>  magic               1.5-9      2018-09-17 [1]
#>  magrittr            2.0.1      2020-11-17 [1]
#>  maps                3.3.0      2018-04-03 [1]
#>  MASS                7.3-54     2021-05-03 [1]
#>  Matrix              1.3-4      2021-06-01 [1]
#>  mclust              5.4.7      2020-11-20 [1]
#>  memoise             2.0.0      2021-01-26 [1]
#>  mFD                 1.0.0      2021-08-17 [1]
#>  mgcv                1.8-36     2021-06-01 [1]
#>  mime                0.11       2021-06-23 [1]
#>  mnormt              2.0.2      2020-09-01 [1]
#>  munsell             0.5.0      2018-06-12 [1]
#>  mvtnorm             1.1-2      2021-06-07 [1]
#>  nlme                3.1-152    2021-02-04 [1]
#>  nls2                0.2        2013-03-08 [1]
#>  numDeriv            2016.8-1.1 2019-06-06 [1]
#>  parallelly          1.27.0     2021-07-19 [1]
#>  patchwork           1.1.1      2020-12-17 [1]
#>  pdist               1.2        2013-02-03 [1]
#>  permute             0.9-5      2019-03-12 [1]
#>  phangorn            2.7.1      2021-07-13 [1]
#>  phylobase           0.8.10     2020-03-01 [1]
#>  phytools            0.7-80     2021-06-03 [1]
#>  pillar              1.6.1      2021-05-16 [1]
#>  pkgconfig           2.0.3      2019-09-22 [1]
#>  pkgdown             1.6.1      2020-09-12 [1]
#>  plotrix             3.8-1      2021-01-21 [1]
#>  plyr                1.8.6      2020-03-03 [1]
#>  png                 0.1-7      2013-12-03 [1]
#>  pracma              2.3.3      2021-01-23 [1]
#>  prettyunits         1.1.1      2020-01-24 [1]
#>  progress            1.2.2      2019-05-16 [1]
#>  promises            1.2.0.1    2021-02-11 [1]
#>  proxy               0.4-26     2021-06-07 [1]
#>  purrr               0.3.4      2020-04-17 [1]
#>  quadprog            1.5-8      2019-11-20 [1]
#>  R6                  2.5.0      2020-10-28 [1]
#>  raster              3.4-13     2021-06-18 [1]
#>  RColorBrewer        1.1-2      2014-12-07 [1]
#>  Rcpp                1.0.7      2021-07-07 [1]
#>  reprex              2.0.0      2021-04-02 [1]
#>  reshape2            1.4.4      2020-04-09 [1]
#>  rgeos               0.5-5      2020-09-07 [1]
#>  rgl                 0.107.10   2021-07-22 [1]
#>  rlang               0.4.11     2021-04-30 [1]
#>  rmarkdown           2.9        2021-06-15 [1]
#>  rncl                0.8.4      2020-02-10 [1]
#>  RNeXML              2.4.5      2020-06-18 [1]
#>  rstudioapi          0.13       2020-11-12 [1]
#>  scales              1.1.1      2020-05-11 [1]
#>  scatterplot3d       0.3-41     2018-03-14 [1]
#>  seqinr              4.2-8      2021-06-09 [1]
#>  sessioninfo         1.1.1      2018-11-05 [1]
#>  sf                  1.0-1      2021-06-29 [1]
#>  shiny               1.6.0      2021-01-25 [1]
#>  sp                  1.4-5      2021-01-10 [1]
#>  spData              0.3.10     2021-06-18 [1]
#>  spdep               1.1-8      2021-05-23 [1]
#>  stringi             1.7.3      2021-07-16 [1]
#>  stringr             1.4.0      2019-02-10 [1]
#>  styler              1.5.1      2021-07-13 [1]
#>  tibble              3.1.3      2021-07-23 [1]
#>  tidyr               1.1.3      2021-03-03 [1]
#>  tidyselect          1.1.1      2021-04-30 [1]
#>  tmvnsim             1.0-2      2016-12-15 [1]
#>  units               0.7-2      2021-06-08 [1]
#>  utf8                1.2.2      2021-07-24 [1]
#>  uuid                0.1-4      2020-02-26 [1]
#>  vctrs               0.3.8      2021-04-29 [1]
#>  vegan               2.5-7      2020-11-28 [1]
#>  viridis             0.6.1      2021-05-11 [1]
#>  viridisLite         0.4.0      2021-04-13 [1]
#>  withr               2.4.2      2021-04-18 [1]
#>  xfun                0.24       2021-06-15 [1]
#>  XML                 3.99-0.6   2021-03-16 [1]
#>  xml2                1.3.2      2020-04-23 [1]
#>  xtable              1.8-4      2019-04-21 [1]
#>  yaml                2.2.1      2020-02-01 [1]
#>  source                            
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.1)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  local                             
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  Github (CmlMagneville/mFD@a6af567)
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.3)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#>  CRAN (R 4.0.2)                    
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

**Comment by Rekyt** [comment]
If I summarize what I found regarding the behavior of function what we want them to output for sites with no species it goes like this:

Function                    | Current Value (on `main`) | Expected Value
---------------------|---------------------------|----------------
`fd_fdis()`                 | `NaN`                                 | 0
`fd_fdiv()`                 | Errors                                  | 0
`fd_feve()`                | Errors                                  | `NA` (for n < 3 species)
`fd_fric_intersect()` | `NA`                                    | `NA` (or 0 ?)
`fd_fric()`                 | Errors                                   | `NA`
`fd_raoq()`                | `NaN`                                 | 0

**Comment by Bisaloo** [comment]
I would return 0 for all indices when there are no species.

FRic, FEve, and intersection of convex hulls have an extra (but distinct) edge where this value is non-zero but cannot be computed (when the number of species is > 0 but lower than what's required to compute it).

Or am I overthinking this?

If you think we should not distinguish the two edge cases, then I'd rather have all functions return the same value for consistency.

**Comment by Rekyt** [comment]
> FRic, FEve, and intersection of convex hulls have an extra (but distinct) edge where this value is non-zero but cannot be computed (when the number of species is > 0 but lower than what's required to compute it).

I 100% agree with you, but it still seems strange to me that for n = 0 we would be able to define these three indices but for 0 < n < n_traits we wouldn't. We can distinguish between both cases, but as shown by other packages, it doesn't seem there is a real consensus solution. I would rather output `NA` for the sake of continuity of the funcitons.

**Comment by Bisaloo** [comment]
What about `NA` for all functions for all edge cases then?

**Comment by Rekyt** [comment]
> What about `NA` for all functions for all edge cases then?

Then it's how it behaves right now :) so everything is good. What do you think of the implementation?

**Comment by Bisaloo** [comment]
Sorry, I mean that FDis, FDiv, and Rao's Q should be NA as well. It feels strange that some functions return 0 and others NA.

File: R/fd_fdiv.R [hunk 2]
```diff
@@ -47,35 +47,42 @@ fd_fdiv <- function(traits, sp_com) {
...
+  sp_com <- sp_com / site_abundances
 
   # Compute Functional Divergence
   fdiv_site <- future_apply(sp_com, 1, function(sp_site) {
 
-    # Select only species that are in site
-    sub_site <- sp_site[sp_site > 0]
+    if (all(is.na(sp_site)) | all(sp_site == 0)) {
+      return(0)
+    } else {
```
Comment (Bisaloo) [inline R/fd_fdiv.R:59]: The `else` is not necessary here because we already have an early `return()` if the `if`.

**Review by Bisaloo — CHANGES_REQUESTED** [review]

**Pull request merged**